File: 424-longest-repeating-character-replacement/longest-repeating-character-replacement.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        myDict = {}
        l = 0
        r = 0
        max_freq = 0
        currSum = 0
        res = 0
        while(r < len(s)):
            if s[r] in myDict:
                myDict[s[r]]+=1
            else:
                myDict[s[r]] = 1
            max_freq = max(myDict.values())
            if r - l + 1 - max_freq > k:
                if myDict[s[l]] == 1:
                    del myDict[s[l]]
                else:
                    myDict[s[l]]-=1
                max_freq = max(myDict.values())
                l+=1
            if r - l + 1 - max_freq <= k:
                res = max(res,r - l + 1)
            r+=1
        return res

File: 424-longest-repeating-character-replacement/longest-repeating-character-replacement.java
class Solution {
    public int characterReplacement(String s, int k) {
        int [] freqs = new int[26];
        int left = 0;
        int right = 0;
        int maxFreq = 0;
        int res = 0;
        while(right < s.length()){
            freqs[s.charAt(right) - 'A']+=1;
            maxFreq = Math.max(maxFreq, freqs[s.charAt(right) - 'A']);
            // if not valid window
            if(right - left + 1 - maxFreq > k){
                freqs[s.charAt(left) - 'A']-=1;
                left+=1;
            }
            // if valid window
            else{
                res = right - left + 1;
            }
            
            right+=1;
        }
        return res;
    }
}